rfcs/2021-08-29-8381-vrl-iteration-support.md

```diff
@@ -0,0 +1,756 @@
+and arrays.
+
+We do this by allowing functions to expose the fact that they accept a closure
+as a stand-alone argument to their function call.
+
+"stand-alone" means the closure comes _after_ the function call itself, e.g.
+this:
+
+```coffee
+map(.) |k, v| {
+  [k, v]
+}
+```
+
+over this:
+
+```coffee
+map(., |k, v| {
+  [k, v]
+})
+```
+
+This choice is made to make it clear that closures in VRL _can't be passed
+around through variables, but are instead syntactically attached to a function
+call_.
+
+That is, we don't want to allow this:
+
+```coffee
+my_closure = |k, v| [k, v]
+map(., my_closure)
+```
+
+As that would require us to (1) allow for a closure to be represented as a `Value`,
+since everything assigned to a variable has to be representable as a value, and
+(2) it would require us to add compiler-support to reject assigning closures to
+an event field (e.g. `.foo = |k, v| [k, v]`), as that would lead to its own set
+of problems.
+
+While closures-as-values might seem useful at first, there hasn't been any
+use-case that has come up in the past that makes this a requirement.
+
+Instead, the closure-syntax is tied to a function call, and can only be added to
+functions that explicitly expose their ability to take a closure with `x`
+arguments that returns `y` value.
+
+#### Lexical Scoping
+
+Lexical scoping (variables being accessible within a given scope, instead of
+globally) is something we've discussed before.
+
+Before, we decided that the complexity of adding lexical scoping wasn't worth
+the investment before our first release, and we also hoped that lexical scoping
+wouldn't be something that was ever needed in VRL.
+
+With this feature, and particular the function-closure syntax, lexical scoping
+comes to top of mind again.
+
+The reason for that, is the following example:
+
+```coffee
+map(.) |key, value| {
+  key = upcase(key)
+
+  [key, value]
+}
+
+key
+```
+
+We reference `key` outside the closure, at the last line of the program. What
+should the value of `key` be in this case?
+
+Without lexical scoping, it would be set to the upper-case variant of the "last"
+key in the event.
+
+With lexical scoping, it would return an "undefined variable" error at
+compile-time, because the `key` variable _inside_ the closure is
+lexically-scoped to that block, and remains undefined outside of the block.
+
+However, while the above syntax would be _new_ and thus not a breaking change,
+for existing code, adding lexical scoping _would_ be a breaking change:
+
+```coffee
+foo = "bar"
+
+{
+  foo = "baz"
+}
+
+foo
+```
+
+Previously, `foo` would return `"baz"` when the program runs, but with lexical
+scoping, it would be set to `"bar"` instead (depending on if we also added
+let-bindings or not).
+
+Because of this breaking change, this RFC proposes we add lexical scoping
+selectively, function-closures _will_ have _partial_ lexical scoping, anything else
+_won't_.
```

Comment:
Agreed. I was a fan initially as well, but worried that this RFC would introduce too many new concepts/breaking changes, and since it wasn't _technically_ required, I removed it in a later version of the RFC.

I'll add it back as something we prefer not to punt on, and solve with this RFC.

Reply:
I can see the concern of not wanting to overload this RFC, but would also very be in favor of having well-defined lexical scoping.

Would it be possible we first do an RFC for lexical scoping, and base this RFC on it? The initial RFC wouldn't need to mention closures yet.

Reply:
@pablosichert's suggestion seems reasonable - especially given we don't _seem_ rushed to land an iteration solution yet, just planning the implementation (I don't believe the work is scheduled in the near term?)

Reply:
Having now added lexical scoping to this RFC in 13ab8eaf45b211991c4aad87e9cbc556a2126457, I don't _think_ it needs its own RFC (the rules are simple), but I wouldn't mind doing the work (after my holiday) of splitting up this RFC into two RFCs if people would prefer that.